File: src/main/java/me/chriss99/Vec4f2DBufferWrapper.java
package me.chriss99;

import org.joml.Vector4f;

import java.nio.ByteBuffer;

import static org.lwjgl.opengl.GL11.*;

public final class Vec4f2DBufferWrapper extends Array2DBufferWrapper {
    public Vec4f2DBufferWrapper(ByteBuffer buffer, int width, int height) {
        super(buffer, Type.VEC4F, width, height);
    }

    public Vec4f2DBufferWrapper(int width, int height) {
        super(Type.VEC4F, width, height);
    }

    public Vec4f2DBufferWrapper mipMap() {
        throw new UnsupportedOperationException("Vec4f can't be mipMapped!");
    }

    public Vector4f getVec(int x, int z) {
        return new Vector4f(
                buffer.getFloat(((z*width + x)*4 + 0)*4),
                buffer.getFloat(((z*width + x)*4 + 1)*4),
                buffer.getFloat(((z*width + x)*4 + 2)*4),
                buffer.getFloat(((z*width + x)*4 + 3)*4));
    }

    public void putVec(int x, int z, Vector4f v) {
        buffer.putFloat(((z*width + x)*4 + 0)*4, v.x);
        buffer.putFloat(((z*width + x)*4 + 1)*4, v.y);
        buffer.putFloat(((z*width + x)*4 + 2)*4, v.z);
        buffer.putFloat(((z*width + x)*4 + 3)*4, v.w);
    }
}

File: src/main/kotlin/Vec4f2DBufferWrapper.kt
package me.chriss99

import org.joml.Vector2i
import org.joml.Vector4f
import java.nio.ByteBuffer

class Vec4f2DBufferWrapper : Array2DBufferWrapper {
    constructor(buffer: ByteBuffer, size: Vector2i) : super(buffer, Type.VEC4F, size)

    constructor(size: Vector2i) : super(Type.VEC4F, size)

    override fun mipMap(): Vec4f2DBufferWrapper? {
        throw UnsupportedOperationException("Vec4f can't be mipMapped!")
    }

    fun getVec(x: Int, z: Int): Vector4f {
        return Vector4f(
            buffer.getFloat(((z * size.x + x) * 4 + 0) * 4),
            buffer.getFloat(((z * size.x + x) * 4 + 1) * 4),
            buffer.getFloat(((z * size.x + x) * 4 + 2) * 4),
            buffer.getFloat(((z * size.x + x) * 4 + 3) * 4)
        )
    }

    fun putVec(x: Int, z: Int, v: Vector4f) {
        buffer.putFloat(((z * size.x + x) * 4 + 0) * 4, v.x)
        buffer.putFloat(((z * size.x + x) * 4 + 1) * 4, v.y)
        buffer.putFloat(((z * size.x + x) * 4 + 2) * 4, v.z)
        buffer.putFloat(((z * size.x + x) * 4 + 3) * 4, v.w)
    }
}